src/equations/compressible_navier_stokes_2d.jl

```diff
@@ -204,6 +208,39 @@ end
   return SVector(rho, v1, v2, T)
 end
 
+
+# Convert conservative variables to entropy
+# Note, only w_2, w_3, w_4 are needed for the viscous fluxes so we avoid computing
+# w_1 and simply copy over rho.
```

Comment:
I think this can hit us hard later if we use something like `cons2entropy` in the initial condition etc. - same problem as with `prim2cons` before. Just something we should be aware of

Reply:
Yeah, I thought about that as well. I was trying to cook-up a way to fix the CNS initial conditions live in the elixir problem but couldn't come up with anything. I believe it all boils down to the `nvars` versus `ngradvars` argument and how to do these different variable conversions.

I am a bit out of my depth because my usual strategy (from my codes with David or FLUXO) is the readin an integer to indicate which variable type I want to use and then use a function pointer for the generic "GetVariblesForGradients" function. I am just not sure how to adapt this into Julia in the best way.

Reply:
We discussed this a bit in our last meeting. One option could be to create a new system of equations for this. Another option could be to let `CompressibleNavierStokesDiffusion2D` contain functions doing this conversion (if we consider it to be part of the physical equations). Yet another option could be to store these functions in the parabolic options (if we consider them to be part of the numerical method). Thoughts, @jlchan @sloede ?

Reply:
Just so I understand: You want to be able to change the variables on which the gradients are computed on the fly, is that correct? But wouldn't you also have to modify the viscous flux function for this to work?

Reply:
I guess the approach he is using right now is to use two functions.

1. Transform the conserved variables to some variables of which you would like to take the gradients.
2. Transform the gradients of the chosen variables to the gradients of some canonical set of variables (here primitive ones, I think).

Then, the viscous fluxes can be written in terms of the gradients of a fixed set of canonical variables.

Reply:
Yes, that is my current strategy that mimics what I have done in the past. If we want to avoid the second `transform_gradients!` step this would require a different
```julia
function flux(u, gradients, orientation::Integer, equations::CompressibleNavierStokesDiffusion2D)
```
where the parabolic terms are directly written in terms of the entropy gradients instead of the primitive variable gradients.

Reply:
OK, thanks for the clarification. But that gives 3 options for any hyp-par system, at least AFAICT: 
* Do everything in conservative variables, only convert ad-hoc in  `flux` as needed (like we do now for hyperbolic stuff)
* Use different set of variables to compute gradients, continue to use those gradients (e.g., primitive)
* Use different set of variables to compute gradients, convert again to different set of gradients (e.g., cons2entropy, entropy2prim)

I think the default behavior should be the first one, that is, if a user implements a new diffusion system, they should not have to consider these conversions at all. Thus whatever strategy we choose, it should keep the code simple when doing simple things.

What is the motivation for it being easy to change the set of variables used to compute the gradients? E.g., would it be ok to just always use entropy vars and that's it? Or is this something that people actively want to experiment with?

If an easy way of changing this needs to be supported, I would prefer a solution that lives in the equations ecosphere. Although it's clearly method-related (if not for stability/numerical reasons, you could just use the conservative variables for the gradients, right?), I think for a user it would be easier to understand if it is closer to where the other, physics-specific details are specified. However, as written before, this would only be suitable to me if it does not require a user implementing a new diffusion system to think of this; if this is not possible, we need to put it with the parabolic options.

Reply:
This warrants some discussion among all of us. I will add it to the agenda for next weeks meeting. But here is an idea. We could make the default to take gradients of the conservative variables (and rewrite the viscous flux accordingly). Then the `transform_variables!` and `transform_gradients!` would "do nothing". But if a user wants to add a variable conversion and gradient conversion these functions then live in the elixir. This is messy but could work. What do you think @sloede @jlchan @ranocha ?

> I would prefer a solution that lives in the equations ecosphere

I agree. This would make the most sense because the different variable types one would want to use and the form of the viscous fluxes are also equation dependent.

> Although it's clearly method-related (if not for stability/numerical reasons, you could just use the conservative variables for the gradients, right?)

Basically yes. You can use the entropy gradients and entropy viscous terms even for an approximation of the "classic" DGSEM, but that would just be a weird combo :) . One just needs to pick a set of gradient variables and then make sure that the viscous flux is written to use the correct set of variables. I selected primitive gradients and the corresponding viscous fluxes to have easier parity with the formulas one sees in papers.

Reply:
> > I would prefer a solution that lives in the equations ecosphere
> 
> I agree. This would make the most sense because the different variable types one would want to use and the form of the viscous fluxes are also equation dependent.

I agree that the implementation should live in the equations folder - but should it be part of the equations struct or of the numerical options, like a numerical flux?

> This warrants some discussion among all of us. I will add it to the agenda for next weeks meeting.

Great, thanks!

Reply:
What happens, if
```julia
cons2gradient(x...) = # convert conservative to gradient variables
gradient2work(x...) = # convert gradient to "working" variables
```
is defined in Trixi and I do
```julia
import Trixi
Trixi.cons2gradient(x...) = # something different
```
in my elixir? Will it get overwritten and "just works", or is it bad style and anything might happen, or maybe it doesn't even work?

I think we are using something similar right now. If this can be formalized, this could be a very easy solution, couldn't it? A downside of this approach, however, is that if you overwrite it once, it will be changed until you restart Julia.

On a separate note: I would like to offer the suggestion that the default variables for the CNS gradients should be the conservative variables. Anything else is imho likely to lead to confusion for new users and would, given that e.g. Euler uses conservative variables, not follow [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Reply:
> Will it get overwritten and "just works", or is it bad style and anything might happen, or maybe it doesn't even work?

It will work but it's not nice since

> it will be changed until you restart Julia.

Reply:
I think the entropy variable formulation should be associated with a new set of equations. We could then transform the gradients inside the flux function and reuse the original CNS flux through `invoke`. 

To me, this would be cleaner than adding a `transform_gradients!` function, since the transformation is independent of the discretization.

Reply:
That's of course another option - and something we agreed to do for now in our last meeting? But maybe it was not clear enough for everybody - let's put it on the agenda for the next meeting

Reply:
> I think the entropy variable formulation should be associated with a new set of equations.

Okay, so I would create another file in the `equations` folder to hold the new formulation for the entropy gradients and Navier-Stokes? I think this strategy will become untenable. I say this because, as a baseline, we would have three versions of the Navier-Stokes diffusion for conservative / primitive / entropy gradients. This would mean another three files for the MHD counterparts. I think this would be very confusing for new users looking into the `equations` folder.

Reply:
I would suggest that we could trim it down to just primitive and entropy variables. The primitive variables are simpler, and the entropy variables are stable. The conservative variables are neither for the parabolic terms

For the complexity, I think the entropy variable equations will actually be pretty short. I can draft a version which reuses the other CNS flux. 

We could also design a single compressible NS equations type which takes in a type parameter to toggle what the gradient variables are. Thoughts?

Reply:
> We could also design a single compressible NS equations type which takes in a type parameter to toggle what the gradient variables are. Thoughts?

That's basically the option to store the transformation in the equations I mentioned above. This would be okay for me.

How many of these different transformations do we really need? Are we likely to come up with new ones (similar to numerical fluxes) or isbthe choice pretty stable?

Reply:
> For the complexity, I think the entropy variable equations will actually be pretty short. I can draft a version which reuses the other CNS flux.

Yeah, for 2D Navier-Stokes (or even 3D I believe) this is not too cumbersome. 
 
> We could also design a single compressible NS equations type which takes in a type parameter to toggle what the gradient variables are. Thoughts?

That could work and be a similar gradient selection strategy that I am used to.

Reply:
> We could also design a single compressible NS equations type which takes in a type parameter to toggle what the gradient variables are.

How would this information be extracted from the equations then without having to implement this for each system of equations?

On a different note: What is the advantage of storing the gradients in primitive variables vs. providing a "gradients_cons2prim" function? I am asking since having to "know" in which system the gradients are given is another potential source of confusion for users (that I actually have experienced myself).

Reply:
> How many of these different transformations do we really need? Are we likely to come up with new ones (similar to numerical fluxes) or is the choice pretty stable?

I've only used primitive and entropy variables. IMO these are pretty stable and established for CNS, but I'm not sure about other systems. 

> How would this information be extracted from the equations then without having to implement this for each system of equations?

I'm not sure if I'm understanding your question. We would have to implement a gradient transformation function for each new set of variables, but we wouldn't have to re-implement everything. 

For example, an implementation for entropy variables (with `EntropyGradientVariables` or `PrimitiveGradientVariables` type parameter added) might look like the following
```julia
function flux(u, gradients, orientation, ::CompressibleNavierStokesDiffusion2D{EntropyGradientVariables})
  gradients_x, gradients_y = gradients
  gradients_x_transformed = convert_gradient_variables(u, gradients_x, equations)
  gradients_y_transformed = convert_gradient_variables(u, gradients_y, equations)

  # this should create a CNS type with the same constants but a `PrimitiveGradientVariables` type parameter
  equations_primitive = CompressibleNavierStokesDiffusion2D{PrimitiveGradientVariables}(equations) 
  return flux(u, gradients_transformed, equations_primitive)

  # we could also use 
  # `invoke(flux, Tuple(typeof{u}, typeof{gradients_transformed}, CompressibleNavierStokesDiffusion{PrimitiveGradientVariables}), u, gradients_transformed, equations)` 
  # to call a specific specialization of `flux`.
end 
```

> On a different note: What is the advantage of storing the gradients in primitive variables vs. providing a "gradients_cons2prim" function? I am asking since having to "know" in which system the gradients are given is another potential source of confusion for users (that I actually have experienced myself).

This wouldn't make a difference for primitive variables, but it would for entropy variables. The formulation isn't provably entropy stable if we compute the conservative variable gradients and then transform them to entropy variables. 

This is because the contribution to the entropy dissipation for entropy variables is `integral( K * (dv/dx)^2)` where `K` is a SPD matrix and `v` are entropy variables. If we used `gradients_cons2prim` instead, we get an entropy dissipation which looks like `integral( (dv/du * du/dx)' * K * (dv/dx))`, which isn't sign-definite.

Reply:
Thanks for the elaborate answer!

> > How would this information be extracted from the equations then without having to implement this for each system of equations?
> 
> I'm not sure if I'm understanding your question. We would have to implement a gradient transformation function for each new set of variables, but we wouldn't have to re-implement everything.

What I meant was: The transformation of the variables before the gradient calculation takes place inside the numerical method. For this to happen, we would have to extract the currently chosen gradient variables from the diffequations object. If we store the variables as a type parameter, I just want to make sure that this will not force us to add this type parameter to *every* diffequations struct, even if only ever one set of variables is used for that system.

> > On a different note: What is the advantage of storing the gradients in primitive variables vs. providing a "gradients_cons2prim" function? I am asking since having to "know" in which system the gradients are given is another potential source of confusion for users (that I actually have experienced myself).
> 
> This wouldn't make a difference for primitive variables, but it would for entropy variables. The formulation isn't provably entropy stable if we compute the conservative variable gradients and then transform them to entropy variables.

Sorry, I think I wasn't clear enough: I am not questioning the need to *compute* the gradients in entropy variables, I am just asking if we need to *store* them in different formats. For instance, we could provide a functionality to allow choosing which variables to use for the gradient calculation, but then to convert these back to the gradients of, e.g., the conservative variables (or the primitive variables). That way, a user would not have to think about which variables the gradients are currently stored in when they write their fluxes etc, since it would always be the same.